directory/directory_dest.go

```diff
@@ -194,23 +195,38 @@ func (d *dirImageDestination) PutBlob(ctx context.Context, stream io.Reader, inp
 // (e.g. if the blob is a filesystem layer, this signifies that the changes it describes need to be applied again when composing a filesystem tree).
 // info.Digest must not be empty.
 // If canSubstitute, TryReusingBlob can use an equivalent equivalent of the desired blob; in that case the returned info may not match the input.
-// If the blob has been successfully reused, returns (true, info, nil); info must contain at least a digest and size.
+// If the blob has been successfully reused, returns (true, info, nil); info must contain at least a digest and size, and may
+// include CompressionOperation and CompressionAlgorithm fields to indicate that a change to the compression type should be
+// reflected in the manifest that will be written.
 // If the transport can not reuse the requested blob, TryReusingBlob returns (false, {}, nil); it returns a non-nil error only on an unexpected failure.
 // May use and/or update cache.
 func (d *dirImageDestination) TryReusingBlob(ctx context.Context, info types.BlobInfo, cache types.BlobInfoCache, canSubstitute bool) (bool, types.BlobInfo, error) {
 	if info.Digest == "" {
 		return false, types.BlobInfo{}, errors.Errorf(`"Can not check for a blob with unknown digest`)
 	}
 	blobPath := d.ref.layerPath(info.Digest)
-	finfo, err := os.Stat(blobPath)
+
+	f, err := os.Open(blobPath)
 	if err != nil && os.IsNotExist(err) {
 		return false, types.BlobInfo{}, nil
 	}
 	if err != nil {
 		return false, types.BlobInfo{}, err
 	}
-	return true, types.BlobInfo{Digest: info.Digest, Size: finfo.Size()}, nil
+	defer f.Close()
+
+	finfo, err := f.Stat()
+	if err != nil {
+		return false, types.BlobInfo{}, err
+	}
+
+	compressionOperation, compressionAlgorithm, _, _, err := blobinfocache.DetectCompressionFormat(f)
```

Comment:
Both in this and the OCI/layout case, the code only looks for exactly the specified digest and no substitution happens. So this should never cause a change in the MIME type, and the two calls to `DetectCompressionFormat` seem unnecessary.

---

OTOH what about the `canSubstitute` case in `storageImageDestination.TryReusingBlob`? That one might need a MIME type update.

Reply:
`storage` is a bit of a weird beast here.  It uses `LayersInfosForCopy()`, so it's never on the hook for being able to reproduce the blobs that it's given.  It's also in the business of trying to avoid causing changes to the manifest even when they're allowed, so it wants to only ever behave as though `canSubstitute` is false.

Reply:
Yes, OTOH we already need to provide an uncompressed size so that the generic copy code is consistent, IIRC, so it makes sense for that rare code path to be fully consistent instead of partially.

Let’s see how the other paths pan out, at this point it’s probably just something to eventually verify.